text/0038-scrubbing-sensitive-data.md

```diff
@@ -0,0 +1,135 @@
+
+The content should not be a simple string but structured data. The content could be a template string where all the sensitive values are removed plus a dictionary with the values to insert into the string.
+Example `span.description` should be a string with named parameters in the format `select * from user where email=%(email)s;` or `POST /api/v1/update_password?new_password=%(new_password)s` and in `span.data` there should be `{"email": "test@example.com"}` or `{ "new_password": "123456" }` respectively.
+Same goes for `breadcrumb.message/breadcrumb.data`, `logentry.message/logentry.params`, and `message.message/message.params`.
+
+We need to identify all the fields we need to do this.
+
+We should model this Option after how `Logentry` does this now.
+
+Note: If we change the span.description then the hash for existing performance issues will be changed and existing performance issues will be recreated, so users would have duplicate performance issues in their list of issues. (Which we can just document in the CHANGELOG and everything should be fine.)
+
+_Pros:_
+
+- Relay would not have to reverse-engineer the semantics of the information supplied by the SDK.
+
+_Cons:_
+
+- Could be complex for nested JSON objects or for "Array of objects" kind of data.
+
+### Option C): Relay identifies what kind of data is present and parses it.
+
+Relay can try for each field to "guess" what kind of data it is. Guessing can be done by looking at what field in general we are guessing (field X has always a SQL query in it), or the span.op or other fields as well as the content itself. If the content is a SQL query, JSON object, Elasticsearch Query, GraphQL Query, URL, ... When Relay is certain that it knows the content is of a specific kind, it can then run a parser on it to be able to scrub values of sensitive fields.
+
+An existing example for this is [parsing URL query parameters into a separate field](https://github.com/getsentry/relay/blob/c2e666d1728a2882b82e70fdbb02192c4cb0b50a/relay-general/src/store/normalize/request.rs#L29-L48), which Relay does when normalizing the [Request Interface](https://develop.sentry.dev/sdk/event-payloads/request/).
+
+For the performance issue detection `sentry` does something similar:
+
+- https://github.com/getsentry/sentry/blob/68e44ed3e8343a5e69d0b0a51ad65c02ae427cd0/src/sentry/spans/grouping/strategy/base.py#L186
+- https://github.com/getsentry/sentry/blob/68e44ed3e8343a5e69d0b0a51ad65c02ae427cd0/src/sentry/spans/grouping/strategy/base.py#L142-L150
+
+_Pros:_
+
+-
+
+_Cons:_
+
+- Could be expensive to try multiple guesses before the right kind of data is identified. (Maybe its SQL? no. Maybe JSON? no. So it is a URL? yes.)
+
+### Option D): Generic tozenization in Relay.
+
+Have a generic tokenizer in Relay that can not parse full fledged SQL, but can extract key/value pairs out of almost everything. With this the values of keys with potential sensitive information can be removed.
```

Comment:
Likely the most expensive one, but it is the one that is implemented on Relay (a single source of truth for data scrubbing), and likely the one that degrades the least the product (scrubbing some key/values rather than the full URL).